lib/ui/painting/image_generator_apng.cc

```diff
@@ -0,0 +1,616 @@
+      case 1:  // APNG_BLEND_OP_OVER
+        frame_info.blend_mode = SkCodecAnimation::Blend::kSrcOver;
+        break;
+      default:
+        return std::make_pair(std::nullopt, nullptr);
+    }
+    switch (control_data->get_dispose_op()) {
+      case 0:  // APNG_DISPOSE_OP_NONE
+        frame_info.disposal_method = SkCodecAnimation::DisposalMethod::kKeep;
+        break;
+      case 1:  // APNG_DISPOSE_OP_BACKGROUND
+        frame_info.disposal_method =
+            SkCodecAnimation::DisposalMethod::kRestoreBGColor;
+        break;
+      case 2:  // APNG_DISPOSE_OP_PREVIOUS
+        frame_info.disposal_method =
+            SkCodecAnimation::DisposalMethod::kRestorePrevious;
+        break;
+      default:
+        return std::make_pair(std::nullopt, nullptr);
+    }
+    uint16_t denominator = control_data->get_delay_den() == 0
+                               ? 100
+                               : control_data->get_delay_den();
+    frame_info.duration =
+        static_cast<int>(control_data->get_delay_num() * 1000.f / denominator);
+
+    result.frame_info = frame_info;
+    result.x_offset = control_data->get_x_offset();
+    result.y_offset = control_data->get_y_offset();
+  }
+
+  std::vector<const ChunkHeader*> image_chunks;
+  size_t chunk_space = 0;
+
+  // Walk the chunks until the next frame, end chunk, or an invalid chunk is
+  // reached, recording the chunks to copy along with their required space.
+  // TODO(bdero): Validate that IDAT/fdAT chunks are contiguous.
+  // TODO(bdero): Validate the acTL/fcTL/fdAT sequence number ordering.
+  do {
+    if (chunk->get_type() != kFrameControlChunkType) {
+      image_chunks.push_back(chunk);
+      chunk_space += GetChunkSize(chunk);
+
+      // fdAT chunks are converted into IDAT chunks when demuxed. The only
+      // difference between these chunk types is that fdAT has a 4 byte
+      // sequence number prepended to its data, so subtract that space from
+      // the buffer.
+      if (chunk->get_type() == kFrameDataChunkType) {
+        chunk_space -= 4;
+      }
+    }
+
+    chunk = GetNextChunk(buffer_p, buffer_size, chunk);
+  } while (chunk != nullptr && chunk->get_type() != kFrameControlChunkType &&
+           chunk->get_type() != kImageTrailerChunkType);
+
+  const uint8_t end_chunk[] = {0,   0,   0,    0,    'I',  'E',
+                               'N', 'D', 0xAE, 0x42, 0x60, 0x82};
+
+  // Form a buffer for the new encoded PNG and copy the chunks in.
+  sk_sp<SkData> new_png_buffer = SkData::MakeUninitialized(
+      header.size() + chunk_space + sizeof(end_chunk));
+
+  {
+    uint8_t* write_cursor =
+        static_cast<uint8_t*>(new_png_buffer->writable_data());
+
+    // Copy the signature/header chunks
+    memcpy(write_cursor, header.data(), header.size());
+    // If this is a frame, override the width/height in the IHDR chunk.
+    if (control_data) {
+      ChunkHeader* ihdr_header =
+          reinterpret_cast<ChunkHeader*>(write_cursor + 8);
```

Comment:
Is the 8 here the size of the PNG signature?

If so, use the constant for the signature length as mentioned above

Reply:
Yeah, done.